packages/compile-solidity/src/compilerSupplier/loadingStrategies/VersionRange.ts

```diff
@@ -124,90 +141,89 @@ export class VersionRange {
   getMostRecentVersionOfCompiler(versions) {
     return versions.reduce((mostRecentVersionFileName, fileName) => {
       const match = fileName.match(/v\d+\.\d+\.\d+.*/);
-      const mostRecentVersionMatch = mostRecentVersionFileName.match(
-        /v\d+\.\d+\.\d+.*/
-      );
+      const mostRecentVersionMatch =
+        mostRecentVersionFileName.match(/v\d+\.\d+\.\d+.*/);
       return semver.gtr(match[0], mostRecentVersionMatch[0])
         ? fileName
         : mostRecentVersionFileName;
     }, "-v0.0.0+commit");
   }
 
-  getSatisfyingVersionFromCache(versionRange) {
+  getSatisfyingVersionFromCache(versionRange: string) {
     if (this.versionIsCached(versionRange)) {
       return this.getCachedSolcByVersionRange(versionRange);
     }
     throw new NoVersionError(versionRange);
   }
 
-  async getAndCacheSolcByUrl(fileName, index = 0) {
-    const url = `${this.config.compilerRoots[index].replace(
-      /\/+$/,
-      ""
-    )}/${fileName}`;
-    const { events } = this.config;
+  async getAndCacheSolcByUrl(fileName: string, index: number) {
+    const { events, compilerRoots } = this.config;
+    const url = `${compilerRoots[index].replace(/\/+$/, "")}/${fileName}`;
     events.emit("downloadCompiler:start", {
       attemptNumber: index + 1
     });
+    let response;
     try {
-      const response = await axios.get(url, { maxRedirects: 50 });
-      events.emit("downloadCompiler:succeed");
-      this.cache.add(response.data, fileName);
-      return this.compilerFromString(response.data);
+      response = await axios.get(url, { maxRedirects: 50 });
     } catch (error) {
       events.emit("downloadCompiler:fail");
-      if (index >= this.config.compilerRoots.length - 1) {
-        throw new NoRequestError("compiler URLs", error);
-      }
-      return this.getAndCacheSolcByUrl(fileName, index + 1);
+      throw error;
     }
+    events.emit("downloadCompiler:succeed");
+    this.cache.add(response.data, fileName);
+    return this.compilerFromString(response.data);
   }
 
-  async getSolcFromCacheOrUrl(versionConstraint) {
-    let allVersions, versionToUse;
-    try {
-      allVersions = await this.getSolcVersions();
-    } catch (error) {
-      throw new NoRequestError(versionConstraint, error);
+  async getSolcFromCacheOrUrl(versionConstraint: string, index: number = 0) {
+    // go through all sources (compilerRoots) trying to locate a
+    // suitable version of the Solidity compiler
+    const { compilerRoots, events } = this.config;
+    if (index > compilerRoots.length - 1) {
+      throw new CompilerFetchingError(compilerRoots);
+    }
+    if (!compilerRoots || compilerRoots.length < 1) {
```

Comment:
The `if (!compilerRoots)` check needs to go before the `index >= compilerRoots.length` check; otherwise an absent `compilerRoots` would cause an error when you attempt to get `compilerRoots.length`.  **This is the reason I'm hitting request changes**.

Also, you don't need a separate case for `compilerRoots.length < 1`; that will be handled automatically by the `index >= compilerRoots.length` check.

Reply:
Yep I ended up switching the order of this as well.

Reply:
...wait, you've got two PRs up at once altering this stuff simultaneously?  Uh, that's confusing... I don't really know how to handle that...

Reply:
With respect to the last point you made in your first comment here, I think we should leave the `compilerRoots.length < 1` check because it yields a different error message. In that case it will tell the user they have no compiler root urls.

Reply:
Makes sense!  I think writing `compilerRoots.length === 0` would be clearer, but it also doesn't really matter. :P

Reply:
agreed!